File: app/src/main/java/com/it/yousefl/dataalgo/sortingalgorithms/SelectionSort.java
package com.it.yousefl.dataalgo.sortingalgorithms;

import java.util.ArrayList;

public class SelectionSort {

    /**
     * ----------------Selection Sort -------------------*
     * <p>
     * it's similar to bubble sort, but instead of first placing large values into sorted position,it
     * places small values into sorted positions.
     */

    public static void main(String[] args) {
        ArrayList<Integer> arrayList = new ArrayList<>();
        arrayList.add(34);
        arrayList.add(22);
        arrayList.add(10);
        arrayList.add(19);
        arrayList.add(17);
        System.out.println(shellSort(arrayList));
    }

    //selectionSort([ 34 , 22 , 10 , 19 , 17])
    //i , j , lowest
    //0 , 1 , 1
    // 0 , 2 , 2
    //  0 , 3 , 2
    // 0 , 4 , 2

    public static ArrayList<Integer> shellSort(ArrayList<Integer> arrayList) {


        for (int i = 0; i < arrayList.size(); i++) {
            int small = i;

            for (int j = i + 1; j < arrayList.size(); j++) {
                if (arrayList.get(j) < arrayList.get(small)) {
                    small = j;
                }
            }
            if (i!=small)
                swap(arrayList, i, small);//
            System.out.println(arrayList);
        }

        return arrayList;
    }

    public static void swap(ArrayList<Integer> arrayList, int large, int small) {
        int temp = arrayList.get(small);
        arrayList.set(small, arrayList.get(large));
        arrayList.set(large, temp);
    }
}

File: app/src/main/java/com/it/yousefl/dataalgo/da/sortingalgorithms/SelectionSort.kt
package com.it.yousefl.dataalgo.da.sortingalgorithms

import java.util.*

/**------------------Selection Sort ------------------*/
fun main() {
    val arrayList = ArrayList<Int>()
    arrayList.add(34)
    arrayList.add(22)
    arrayList.add(10)
    arrayList.add(19)
    arrayList.add(17)
    println(shellSort(arrayList))
}

//selectionSort([ 34 , 22 , 10 , 19 , 17])
//i , j , lowest
//0 , 1 , 1
// 0 , 2 , 2
//  0 , 3 , 2
// 0 , 4 , 2

//selectionSort([ 34 , 22 , 10 , 19 , 17])
//i , j , lowest
//0 , 1 , 1
// 0 , 2 , 2
//  0 , 3 , 2
// 0 , 4 , 2
fun shellSort(arrayList: ArrayList<Int>): ArrayList<Int>? {
    for (i in arrayList.indices) {
        var small = i
        for (j in i + 1 until arrayList.size) {
            if (arrayList[j] < arrayList[small]) {
                small = j
            }
        }
        if (i != small) swapSelection(arrayList, i, small) //
        println(arrayList)
    }
    return arrayList
}

fun swapSelection(arrayList: ArrayList<Int>, large: Int, small: Int) {
    val temp = arrayList[small]
    arrayList[small] = arrayList[large]
    arrayList[large] = temp
}